File: Loenn/entities/coreZone.lua
local mods = require("mods")

local styleMaskHelperLoaded = mods.hasLoadedMod("StyleMaskHelper")

local coreZone = {
    name = "EeveeHelper/CoreZone",
    depth = 10000,
    fillColor = { 0.2, 1.0, 1.0, 0.4 },
    borderColor = { 0.2, 1.0, 1.0, 1.0 },

    placements = {
        default = {
            data = {
                width = 8,
                height = 8,
                zoneId = "",
                coreMode = "Current",
                colorGradeMask = false,
                stylegroundMask = false
            }
        },
        {
            name = "current",
            data = { coreMode = "Current", width = 8, height = 8 }
        },
        {
            name = "inverted",
            data = { coreMode = "Inverted", width = 8, height = 8 }
        },
        {
            name = "hot",
            data = { coreMode = "Hot", width = 8, height = 8  }
        },
        {
            name = "cold",
            data = { coreMode = "Cold", width = 8, height = 8  }
        }
    },

    fieldInformation = {
        coreMode = {
            options = { "Current", "Inverted", "Hot", "Cold", "None" },
            editable = false
        }
    },

    ignoredFields = function (entity)
        local ignored = {"_name", "_id", "originX", "originY"}

        if not styleMaskHelperLoaded then
            if not entity.colorGradeMask then
                table.insert(ignored, "colorGradeMask")
            end
            if not entity.stylegroundMask then
                table.insert(ignored, "stylegroundMask")
            end
        end

        return ignored
    end,

    associatedMods = function (entity)
        if entity.colorGradeMask or entity.stylegroundMask then
            return {"EeveeHelper", "StyleMaskHelper"}
        else
            return {"EeveeHelper"}
        end
    end
}

return coreZone

File: Ahorn/entities/coreZone.jl
module EeveeHelperCoreZone

using ..Ahorn, Maple

@mapdef Entity "EeveeHelper/CoreZone" CoreZone(x::Integer, y::Integer, width::Integer=8, height::Integer=8,
    zoneId::String="", coreMode::String="Current", colorGradeMask::Bool=false, stylegroundMask::Bool=false)

const placements = Ahorn.PlacementDict(
    "Core Zone (Eevee Helper)" => Ahorn.EntityPlacement(
        CoreZone,
        "rectangle"
    )
)

Ahorn.minimumSize(entity::CoreZone) = 8, 8
Ahorn.resizable(entity::CoreZone) = true, true

Ahorn.editingOptions(entity::CoreZone) = Dict{String, Any}(
    "coreMode" => ["Current", "Inverted", "Hot", "Cold", "None"]
)

Ahorn.selection(entity::CoreZone) = Ahorn.getEntityRectangle(entity)

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::CoreZone, room::Maple.Room)
    width = Int(get(entity.data, "width", 32))
    height = Int(get(entity.data, "height", 32))

    Ahorn.drawRectangle(ctx, 0, 0, width, height, (0.2, 1.0, 1.0, 0.4), (0.2, 1.0, 1.0, 1.0))
end

end